CanonicalABI.md

```diff
@@ -0,0 +1,976 @@
+  signature with interface types and the lowering operation is done from
+  interface types values to core wasm ABI values.
+
+* Lifting an imported function call - this is done on the opposite side of an
+  imported function. This could be either a host or a different WebAssembly
+  module. The operations here are defined by transforming core wasm values into
+  interface values.
+
+* Lowering an exported function call - this means that an external host or
+  external WebAssembly module is calling an exported function of a WebAssembly
+  module. Here interface types values, matching the type signature of the
+  module's export, are lowered into core wasm values.
+
+* Lifting an exported function call - this is where a WebAssembly module's
+  export converts the core wasm arguments it received into interface types
+  values.
+
+With a canonical ABI it's expected that almost all of the above operations will
+largely be handled by code generators in various languages. For example if
+you're compiling a Rust project to WebAssembly you'd use a code generator to
+generate all the glue necessary to convert from Rust values to core wasm values,
+according to the canonical ABI.
+
+An expected example of the canonical ABI will be WASI and the functions that it
+supports. The vision for how this might looks is:
+
+1. WASI is first defined in terms of interface types. This means all of its
+   functions will be only using types from the interface types specification.
+
+2. When a WebAssembly module, say C or Rust, calls a WASI API then a code
+   generator for the appropriate language will be used to have an API, in that
+   language, which takes types of that language (e.g. Rust structs, C unions,
+   etc) and then converts those values to the canonical ABI (according to this
+   document). The WebAssembly will have a core wasm import of a WASI function
+   which is expected to behave according to this document.
+
+3. On the implementor side of things, a JS host, for example, would use a code
+   generator to generate an object where it takes core wasm values and then
+   hooks up to a JS implementation of WASI. The JS implementation would only use
+   JS objects and wouldn't need to interact with the wasm module/instance itself
+   (due to the shared-nothing nature of interface types).
+
+Eventually with a binary encoding of interface types it can be imagined that
+WebAssembly modules are wrapped in a thin binary encoding which ascribes the
+types of all imports and exports. With these types various interface types
+signatures can be checked at instantiation time to ensure that the stars all
+align and the ABIs are expected to match. In the very immediate future, however,
+a binary encoding is optional because it can simply be understood that a wasm
+module's imports/exports are described with interface types and both sides of
+the boundary are using the same interface types to generate glue code.
```

Comment:
I'd leave out this paragraph.  While *this* doc doesn't define the binary encoding, another will, so it's perhaps better just to leave it open in this doc how exactly these types "show up" in a module/component.